docs/design/2021-03-09-dynamic-privileges.md

```diff
@@ -0,0 +1,348 @@
+| `RESTRICTED_TABLES` | A special privilege which means that the SEM hidden table semantic doesn’t apply. | It is intended for the CloudAdmin user in DbaaS. | 
+
+### Parser Changes
+
+The parser already supports `DYNAMIC` privileges. This can be confirmed by the following patch to TiDB, where they are sent as the static type of `ExtendedPriv`:
+
+```
+diff --git a/planner/core/planbuilder.go b/planner/core/planbuilder.go
+index 90d5b9e82..d1644ee83 100644
+--- a/planner/core/planbuilder.go
++++ b/planner/core/planbuilder.go
+@@ -2298,6 +2298,11 @@ func collectVisitInfoFromGrantStmt(sctx sessionctx.Context, vi []visitInfo, stmt
+ 
+        var allPrivs []mysql.PrivilegeType
+        for _, item := range stmt.Privs {
++
++               if item.Priv == mysql.ExtendedPriv {
++                       fmt.Printf("### Attempting to set DYNAMIC privilege: %s\n", item.Name)
++               }
++
+                if item.Priv == mysql.AllPriv {
+                        switch stmt.Level.Level {
+                        case ast.GrantLevelGlobal:
+```
+
+This results in the following written to the log file:
+```
+mysql> grant acdc on *.* to u1;
+ERROR 8121 (HY000): privilege check fail
+
+### Attempting to set DYNAMIC privilege: acdc
+```
+
+It might be possible that changes are still required if `ExtendedPriv` is not supported in all contexts (REVOKE, etc).
+
+Note that creating a role with the same name as a DYNAMIC privilege is supported. A `GRANT` statement can be attributed to ROLES when it omits the ON *.* syntax. Thus:
+
+```
+GRANT BINLOG_ADMIN TO u1; // grants the role binlog_admin
+GRANT BINLOG_ADMIN ON *.* TO u1; // grants the dynamic privilege binlog_admin
+```
+
+This same nuance applies to MySQL.
+
+### Documentation Plan
+
+The statement reference pages for each of the affected metadata commands will need to be updated to describe dynamic privileges.
+
+There will also need to be documentation specific to `DYNAMIC` privileges to describe how it works, and the purpose of fine-grained access control.
+
+## Test Design
+
+Testing dynamic privileges is a little bit complex because of the various ways privileges can be inherited by users:
+
+* Direct `GRANT` to the user
+* Granting to a role that the user inherits.
+
+### Functional Tests
+
+Unit tests will be added to cover the semantics around role/dynamic privilege precedence, including logical restoring in a different order.
+
+Unit tests will also cover each of the "initial set of Dynamic privileges". Tests will include both directly assigning the privileges and assigning via a ROLE (RBAC).
+
+Integration testing needs to test with global kill enabled/disabled.
+
+### Scenario Tests
+
+The use-cases required by the DBaaS team should be validated when combined with `security-enhanced-mode`. They are:
+
+| Account Name | root | cloudAdmin |
+| --------------- | --------------- | --------------- |
+| Backup & Restore to cloud | Y | Y |
+| File privilege | N | N |
```

Comment:
Another question is currently `cloudAdmin` does not have FILE privilege, but if SREs do want this privilege in the future, how will they grant it to themselves.

Reply:
It is possible by starting the server up in `skip-grant-tables` mode. I would say for security reasons they should avoid needing `FILE` though. Instead of using `SELECT INTO OUTFILE`, they can use the client to write the file:
```bash
mysql -BNe "SELECT * FROM db.table" > output.txt
```

This is a lot less vulnerable to privilege escalations.

Reply:
Can `cloudAdmin` insert into some `information_schema` table to obtain some privileges?

I just point FILE as an example, in the future we may implement more dynamic privileges and SRE may want to use some of them.

Reply:
Information_schema is always non-insertable. I think we would have to write a `session/bootstrap.go` task to essentially "split" a privilege to say if you have `XYZ` now you have `XYZ` and `ZYX`.

The problem with allowing cloudAdmin to extend their privileges is that it is essentially a privilege escalation, since you are only supposed to be able to grant privileges that you already have. There is one such "legal" privilege escalation that is hard coded into this design (`SUPER` always satisfies a non-restricted dynamic privilege). The design of this legal escalation is to get around a limitation:

`visitorInfo` does not currently support the concept of an `OR` for permissions, so while in MySQL it is possible to require either `BACKUP_ADMIN` or `SUPER` the actual implementation in TiDB is that it requires `BACKUP_ADMIN`, but the privilege check API will explicitly fallback and check if the `SUPER` privilege is assigned.

Reply:
sorry, I said the wrong table before. cloudAdmin can insert `mysql.user` table to extend their privilege by `GRANT SELECT, INSERT, UPDATE ON mysql.* TO 'cloudAdmin'@'%%';` when initializing the account.

I know there's no design for privilege escalation so maybe inserting into `mysql.user` is the best solution.
And I'm happy if you put this solution into your design doc and future user doc.

Reply:
It would be `mysql.user` to extend static privileges or `mysql.global_grants` for `XYZ` since it is a dynamic priv, followed by `FLUSH PRIVILEGES`. I can add a section on extending privileges; it's actually a risk to grant `SELECT, INSERT, UPDATE ON mysql.*`

Reply:
This has been added to unanswered questions here: https://github.com/pingcap/tidb/pull/23224/commits/0ea98a17d4e3dbd45b02b95675e8e66767f1e2e7

Reply:
I think "split" a privilege is too hacking in bootstrap.go. If you like this option, I say why not we maintain the cloudAdmin account in bootstrap.go